File: src/RAM.java
import java.io.File;
import java.io.FileInputStream;
import java.nio.ByteBuffer;

public class RAM {
    private final ByteBuffer memory;

    public RAM(int size) {
        memory = ByteBuffer.allocate(size);
    }

    public void write(int location, byte data) {
        if(location >= memory.limit()) {
            throw new ArrayIndexOutOfBoundsException("Ram is only " + memory.limit() + "bytes long");
        }
        memory.put(location, data);
    }

    public void writeBytes(int location, ByteBuffer data, int offset, int size) {
        if(location + size > memory.limit()) {
            throw new ArrayIndexOutOfBoundsException("Ram is only " + Settings.RAM_SIZE + "bytes long");
        }
        memory.put(location, data, offset, size);
    }

    public byte read8(int location) {
        if(location >= memory.limit()) {
            throw new ArrayIndexOutOfBoundsException("Ram is only " + memory.limit() + "bytes long");
        }
        return memory.get(location);
    }
    public short read16(int location) {
        if (location+1 >= memory.limit()){
            throw new ArrayIndexOutOfBoundsException("Ram is only " + memory.limit() + "bytes long");
        }
        return memory.getShort(location);
    }

    public ByteBuffer readBytes(int location, int size) {
        if(location + size >= memory.limit()) {
            throw new ArrayIndexOutOfBoundsException("Ram is only " + memory.limit() + "bytes long");
        }
        return memory.slice(location, size);
    }

    public void loadROMFile(int location, File file) {
        try (FileInputStream inputStream = new FileInputStream(file)) {
            var rom = inputStream.readAllBytes();
            writeBytes(location, ByteBuffer.wrap(rom), 0, rom.length);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: makkan-8/RAM.cs
using System.ComponentModel.DataAnnotations;
using System.Diagnostics;

namespace makkan_8;

public class RAM
{
    private byte[] memory;
    private MemoryStream stream;

    public RAM(int size)
    {
        memory = new byte[size];
        stream = new MemoryStream(memory);
        memory[0] = 0b1111;
        memory[1] = 0b1111;
    }
    public void Write(int location, byte data)
    {
        if (location >= memory.Length)
        {
            throw new IndexOutOfRangeException($"Ram is only {memory.Length} long!");
        }
        memory[location] = data;
    }

    public void WriteBytes(int location, MemoryStream stream, int offset, int size)
    {
        if (location >= memory.Length)
        {
            throw new IndexOutOfRangeException($"Ram is only {memory.Length} long!");
        }

        this.stream.Position = location;
        this.stream.Write(stream.ToArray(), offset, size);
    }

    public byte Read8(int location)
    {
        if (location >= memory.Length)
        {
            throw new IndexOutOfRangeException($"Ram is only {memory.Length} long!");
        }

        return memory[location];
    }

    public ushort Read16(int location)
    {
        if (location+1 >= memory.Length)
        {
            throw new IndexOutOfRangeException($"Ram is only {memory.Length} long!");
        }
        
        return (ushort) ((memory[location] << 8) | memory[location + 1]);
    }
    
    public void LoadRomFile(int location, String filepath)
    {
        if (!File.Exists(filepath)) throw new FileNotFoundException($"File at {Path.GetFullPath($"./{filepath}")} doesn't exist.");
        try
        {
            var rom = File.ReadAllBytes(filepath);
            Console.Out.Write($"ROM read-in size: {rom.Length}");
            WriteBytes(location, new MemoryStream(rom), 0, rom.Length);
        }
        catch (Exception e)
        {
            Debug.WriteLine(e.Message);
        }
    }

}